### ISSUE TITLE
NoMethodError on a virtual collection of POROs

### ISSUE THREAD

**Issue by svoop** [trailblazer/reform#426]
Here's an example similar to the "uninitialized collections" example on the TRB site:

```ruby
class Song
  attr_accessor :title
end

class Album
  attr_accessor :name
end

class AlbumForm < Reform::Form
  property :name
  collection :songs, virtual: true, populate_if_empty: Song do
    property :title
  end
end

album = Album.new
album.songs = []     # <-- same error below no matter whether this line is present or not
form  = AlbumForm.new(album)

form.validate(name: "album-name", songs: [{title: "song-title"}])
# => NoMethodError: undefined method `original' for nil:NilClass
```

As a workaround, the initialization can be made on the form itself after the form is initialized:

```ruby
form.songs = []
```

However, there's a second maybe unrelated issue. With the above workaround in place the following happens:

```ruby
form.validate(name: "album-name", songs: [{title: "song-title"}])
# => true
form.save
# => NoMethodError: undefined method `save' for Song
```

Adding an empty `save` method to the `Song` class serves as a workaround, but since the collection is virtual, why is the `save` message sent to the song objects in the first place?

**Comment by svoop** [comment]
There's `save: false` to prevent the latter problem, I was expecting `virtual: true` to implicitly prevent saving.

**Comment by fran-worley** [comment]
I think you shouldn't need to override the initialiser method if you add a songs method to your albums class that by default returns an empty array.

I think the reason it's calling save is because you can't make an entire collection virtual, only individual properties.

Can you explain a little more of your use case so we can see if it's a feature worth adding?

**Comment by svoop** [comment]
My use case is a little complicated since we're working in a heavily multi-lingual environment here. 

However, it boils down to two model attributes holding arrays of strings, one with subtitles and the other with the content following the subtitle. The form is supposed to group those as an array of subtitle/content tuples in order to always add or remove subtitles and their content together.

So indeed, I was expecting the collection (of subtitle and content tuples) to be virtual by declaring it so.

**Comment by lastobelus** [comment]
This error really threw me -- in particular because if you **don't** use a populator, it works & the virtual collection gets populated with the incoming hashes. However, this error also occurs if you don't specify a populator, but use a `form:` option on the collection. It's very much a trap for the unwary, because looking at the stack traces doesn't really provide any insight to what's going on.

Example:

**Works** 

    class VendorOrderForm < Reform::Form
      property :id

      collection :items, virtual: true
    end


**Asplodes**

    class VendorLineItemForm < Reform::Form
      property :line_item_id
    end

    class VendorOrderForm < Reform::Form
      property :id
      collection :items, virtual: true, form: VendorLineItemForm
    end

...but works if you do `form.items = []; form.validate(...)`

@fran-worley why should a **virtual** collection `songs` call songs on `Album`? Do you mean to add `def songs;[];end` to `AlbumForm`? When I tried that I got `undefined method 'original' for []:Array`. In any case, I did try adding `def items;[];end` to the underlying model for VendorOrderForm in my example, and still got `undefined method 'original' for nil:NilClass`

**Comment by mpospelov** [comment]
For my case case I do something likee this
```
collection :items, form: VendorLineItemForm, populate_if_empty: OpenStruct, virtual: true, default: []
```
and it stops raising this error

**Comment by emaglio** [comment]
This should be fixed with #485
